File: scala/problem-00003.scala
/*
The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?
*/
import scala.Math._

def fermatFactors(n: Long): Set[Long] = {
  if (n % 2 == 0) return Set(2, n / 2) 
  var a = ceil(sqrt(n))
  var b = a * a - n
  var s = sqrt(b)
  while (s != round(s)) {
    a += 1
    b = a * a -n
    s = sqrt(b)
  }
  Set(a - s, a + s).map(_.toLong)
}

def primeFactors(n: Long): Set[Long] = {
  val factors = fermatFactors(n)
  if (factors contains 1) 
    factors - 1
  else 
    factors.map(f => primeFactors(f)).reduceLeft(_ ++ _)
}

val result = primeFactors(600851475143L).reduceLeft(_ max _)
println(result)

File: haskell/problem-00003.hs
{-
The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?

[minhuang@mocha:haskell]$ time runghc problem-00003.hs 
6857

real    0m0.314s
user    0m0.280s
sys     0m0.032s
-}
import Data.List

f :: Integer -> Integer
f x = x^2+1

factors' :: Integer -> Integer -> Integer -> Integer -> Integer
factors' x y d n | d > 1 = d
factors' x y d n = factors' a b (gcd c n) n
    where a = mod (f x) n
          b = mod (f . f $ y) n
          c = abs (a-b)

factors :: Integer -> [Integer]
factors n = [factor, n `div` factor]
    where factor = factors' 2 2 1 n

primeFactors :: Integer -> [Integer]
primeFactors n = case factors n of
    xs | elem 1 xs -> xs
    xs -> nub . foldr1 (++) . map primeFactors $ xs

result = maximum . primeFactors $ 600851475143
main = print result
